File: src/rss.ts
import type { ParseOptions } from './scraper'

export const RSS_OPTIONS: ParseOptions = {
	linkExpr: '/rss/channel/link/text()',
	titleExpr: '/rss/channel/title/text()',
	entriesExpr: '/rss/channel/item',
	entryLinkExpr: '/link/text()',
	entryTitleExpr: '/title/text()',
	entryPublishedExpr: '/pubDate/text()',
	entryDescriptionExpr: '/description/text()',
	entryAuthorExpr: '/author/text()',
	entryThumbnailExpr: "/enclosure[starts-with(@type, 'image/')]/@url",
}

File: crates/scraper/src/feed/rss.rs
use colette_feed::rss::{RssFeed, RssItem};

use super::{ExtractedFeed, ExtractedFeedEntry};

impl From<RssFeed> for ExtractedFeed {
    fn from(value: RssFeed) -> Self {
        Self {
            link: Some(value.channel.link),
            title: Some(value.channel.title),
            entries: value
                .channel
                .item
                .into_iter()
                .map(ExtractedFeedEntry::from)
                .collect(),
        }
    }
}

impl From<RssItem> for ExtractedFeedEntry {
    fn from(value: RssItem) -> Self {
        Self {
            link: Some(value.link),
            title: Some(value.title),
            published: value.pub_date,
            description: Some(value.description),
            author: value.author,
            thumbnail: None,
        }
    }
}
